File: packages/skillet-eval-python/src/skill_eval/runner/agent_loop.py
from __future__ import annotations

import json
import os
from collections.abc import Awaitable, Callable

from skill_eval.providers.types import (
    ChatParams,
    ChatResponse,
    LLMProvider,
    Message,
    ToolDefinition,
    ToolHandler,
)
from skill_eval.utils.async_utils import with_timeout
from skill_eval.utils.error import extract_error_message
from skill_eval.utils.string import truncate

from .context import (
    compact_messages,
    debug_payload_log,
    extract_snippet,
    summarize_tool_args,
    total_payload_size,
)
from .transcript import ToolResult, TranscriptStep, build_agent_run

CHARS_PER_TOKEN = 4
API_CALL_TIMEOUT_SEC = 120


def _usage_dict(response: ChatResponse) -> dict[str, int]:
    u = response.usage
    if hasattr(u, "input_tokens"):
        return {"inputTokens": u.input_tokens, "outputTokens": u.output_tokens}
    return {
        "inputTokens": u.get("inputTokens", u.get("input_tokens", 0)),
        "outputTokens": u.get("outputTokens", u.get("output_tokens", 0)),
    }


async def run_agent_loop(
    *,
    provider: LLMProvider,
    system: str,
    user_prompt: str | None = None,
    turns: list[str] | None = None,
    tools: list[ToolDefinition],
    tool_handlers: dict[str, ToolHandler],
    max_steps: int = 20,
    temperature: float | None = None,
    on_activity: Callable[[str], None] | None = None,
    check_turn_relevance: Callable[[str, str], Awaitable[bool]] | None = None,
):
    resolved_turns = turns if turns is not None else ([user_prompt] if user_prompt else [])
    if not resolved_turns:
        raise ValueError("No user turns provided")

    next_turn = 1
    current_turn = 0
    emit = on_activity or (lambda _: None)
    transcript: list[TranscriptStep] = []
    messages: list[Message] = [Message(role="user", content=resolved_turns[0])]
    debug = bool(os.environ.get("SKILL_EVAL_DEBUG"))

    for step in range(max_steps):
        if debug:
            pre_size, breakdown = debug_payload_log(step, messages, system)
            print(
                f"\n[debug] step {step + 1} | {len(messages)} msgs | "
                f"~{pre_size / 1024:.0f}KB (~{round(pre_size / CHARS_PER_TOKEN):,} tok)",
                flush=True,
            )
            for line in breakdown:
                print(line, flush=True)

        emit(f"step {step + 1} — calling model…")

        pre_size = total_payload_size(system, messages) if debug else 0
        compact_messages(messages, system)

        if debug:
            post_size = total_payload_size(system, messages)
            if post_size < pre_size:
                print(
                    f"[debug] compacted {pre_size / 1024:.0f}KB → {post_size / 1024:.0f}KB",
                    flush=True,
                )

        response = await with_timeout(
            provider.chat(ChatParams(system=system, messages=messages, tools=tools, temperature=temperature)),
            API_CALL_TIMEOUT_SEC,
            f"provider.chat (step {step + 1})",
        )

        transcript_step = TranscriptStep(
            step=step,
            turn=current_turn,
            response=response.content,
            tool_calls=response.tool_calls,
            tool_results=[],
            usage=_usage_dict(response),
            latency_ms=response.latency_ms,
        )

        text_snippet = extract_snippet(response.content)
        if text_snippet:
            emit(text_snippet)

        if response.stop_reason != "tool_use" or not response.tool_calls:
            transcript.append(transcript_step)

            if next_turn < len(resolved_turns):
                user_reply = resolved_turns[next_turn]

                if check_turn_relevance:
                    emit("checking turn relevance…")
                    relevant = await check_turn_relevance(response.content, user_reply)
                    if not relevant:
                        emit(
                            "turn mismatch — agent didn't ask for expected input, ending conversation"
                        )
                        transcript_step.turn_skipped = user_reply
                        return build_agent_run(transcript)

                messages.append(
                    Message(role="assistant", content=response.content)
                )
                current_turn = next_turn
                next_turn += 1
                emit(f"turn {current_turn + 1} — user: {truncate(user_reply, 60)}")
                transcript_step.user_message = user_reply
                messages.append(Message(role="user", content=user_reply))
                continue

            emit(f"done — {text_snippet}" if text_snippet else "done")
            return build_agent_run(transcript)

        messages.append(
            Message(
                role="assistant",
                content=response.content,
                tool_calls=response.tool_calls,
            )
        )

        for tc in response.tool_calls or []:
            arg_snippet = summarize_tool_args(tc.name, tc.arguments)
            emit(f"{tc.name}({arg_snippet})")

            handler = tool_handlers.get(tc.name)
            result: object
            if handler:
                try:
                    result = await handler(tc.arguments)
                except Exception as err:
                    result = {"error": extract_error_message(err)}
            else:
                result = {"error": f"Unknown tool: {tc.name}"}

            transcript_step.tool_results.append(
                ToolResult(id=tc.id, name=tc.name, result=result)
            )
            messages.append(
                Message(
                    role="tool_result",
                    content=json.dumps(result, default=str),
                    tool_call_id=tc.id,
                )
            )

        transcript.append(transcript_step)

    return build_agent_run(transcript)

File: packages/skillet/src/eval/runner/agent-loop.ts
import type { LLMProvider, Message, ToolDefinition, ToolHandler } from "../providers/types.js";
import { withHeartbeat, withTimeout } from "../utils/async.js";
import { extractErrorMessage } from "../utils/error.js";
import { truncate } from "../utils/string.js";
import {
	compactMessages,
	debugPayloadLog,
	extractSnippet,
	summarizeToolArgs,
	totalPayloadSize,
} from "./context.js";
import type { TranscriptStep } from "./transcript.js";
import { type AgentRun, buildAgentRun } from "./transcript.js";

const CHARS_PER_TOKEN = 4;
const API_CALL_TIMEOUT_MS = 120_000;
const HEARTBEAT_INTERVAL_MS = 15_000;

function formatElapsed(ms: number): string {
	const seconds = Math.floor(ms / 1000);
	if (seconds < 60) return `${seconds}s`;
	const minutes = Math.floor(seconds / 60);
	return `${minutes}m${seconds % 60}s`;
}

export interface AgentLoopParams {
	provider: LLMProvider;
	system: string;
	/** Single-turn prompt (deprecated — prefer `turns`). */
	userPrompt?: string;
	/** Ordered user messages. First is sent immediately; subsequent ones are
	 *  injected each time the model yields (responds without tool calls). */
	turns?: string[];
	tools: ToolDefinition[];
	toolHandlers: Record<string, ToolHandler>;
	maxSteps?: number;
	temperature?: number;
	onActivity?: (detail: string) => void;
	heartbeatIntervalMs?: number;
	/** When provided, called before injecting the next scripted turn to verify
	 *  the reply makes sense given the agent's last response. Return false to
	 *  stop injecting turns (conversation ends early). */
	checkTurnRelevance?: (agentResponse: string, nextUserMessage: string) => Promise<boolean>;
}

export async function runAgentLoop(params: AgentLoopParams): Promise<AgentRun> {
	const {
		provider,
		system,
		tools,
		toolHandlers,
		maxSteps = 20,
		temperature,
		onActivity,
		heartbeatIntervalMs = HEARTBEAT_INTERVAL_MS,
	} = params;

	const turns = params.turns ?? (params.userPrompt ? [params.userPrompt] : []);
	if (!turns.length) throw new Error("No user turns provided");

	let nextTurn = 1;
	let currentTurn = 0;
	const emit = onActivity ?? (() => {});
	const transcript: TranscriptStep[] = [];
	const messages: Message[] = [{ role: "user", content: turns[0] }];

	const debug = !!process.env.SKILL_EVAL_DEBUG;

	for (let step = 0; step < maxSteps; step++) {
		if (debug) {
			const { preSize, breakdown } = debugPayloadLog(step, messages, system);
			console.error(
				`\n[debug] step ${step + 1} | ${messages.length} msgs | ~${(preSize / 1024).toFixed(0)}KB (~${Math.round(preSize / CHARS_PER_TOKEN).toLocaleString()} tok)`,
			);
			if (breakdown.length) console.error(breakdown.join("\n"));
		}

		emit(`step ${step + 1} — calling model…`);

		const preSize = debug ? totalPayloadSize(system, messages) : 0;
		compactMessages(messages, system);

		if (debug) {
			const postSize = totalPayloadSize(system, messages);
			if (postSize < preSize) {
				console.error(
					`[debug] compacted ${(preSize / 1024).toFixed(0)}KB → ${(postSize / 1024).toFixed(0)}KB`,
				);
			}
		}

		const response = await withHeartbeat(
			withTimeout(
				provider.chat({ system, messages, tools, temperature }),
				API_CALL_TIMEOUT_MS,
				`provider.chat (step ${step + 1})`,
			),
			{
				intervalMs: heartbeatIntervalMs,
				onHeartbeat: (elapsedMs) =>
					emit(`step ${step + 1} — still waiting on model… ${formatElapsed(elapsedMs)}`),
			},
		);

		const transcriptStep: TranscriptStep = {
			step,
			turn: currentTurn,
			response: response.content,
			toolCalls: response.toolCalls,
			toolResults: [],
			usage: response.usage,
			latencyMs: response.latencyMs,
		};

		const textSnippet = extractSnippet(response.content);
		if (textSnippet) {
			emit(textSnippet);
		}

		if (response.stopReason !== "tool_use" || !response.toolCalls?.length) {
			transcript.push(transcriptStep);

			if (nextTurn < turns.length) {
				const userReply = turns[nextTurn];

				if (params.checkTurnRelevance) {
					emit("checking turn relevance…");
					const relevant = await params.checkTurnRelevance(response.content, userReply);
					if (!relevant) {
						emit("turn mismatch — agent didn't ask for expected input, ending conversation");
						transcriptStep.turnSkipped = userReply;
						return buildAgentRun(transcript);
					}
				}

				messages.push({ role: "assistant", content: response.content });
				currentTurn = nextTurn;
				nextTurn++;
				emit(`turn ${currentTurn + 1} — user: ${truncate(userReply, 60)}`);
				transcriptStep.userMessage = userReply;
				messages.push({ role: "user", content: userReply });
				continue;
			}

			emit(textSnippet ? `done — ${textSnippet}` : "done");
			return buildAgentRun(transcript);
		}

		messages.push({
			role: "assistant",
			content: response.content,
			toolCalls: response.toolCalls,
			_rawParts: response._rawParts,
		});

		for (const tc of response.toolCalls) {
			const argSnippet = summarizeToolArgs(tc.name, tc.arguments);
			emit(`${tc.name}(${argSnippet})`);

			const handler = toolHandlers[tc.name];
			let result: unknown;
			if (handler) {
				try {
					result = await handler(tc.arguments);
				} catch (err) {
					result = { error: extractErrorMessage(err) };
				}
			} else {
				result = { error: `Unknown tool: ${tc.name}` };
			}

			transcriptStep.toolResults.push({ id: tc.id, name: tc.name, result });
			messages.push({
				role: "tool_result",
				content: JSON.stringify(result),
				toolCallId: tc.id,
			});
		}

		transcript.push(transcriptStep);
	}

	return buildAgentRun(transcript);
}
